src/box/iproto.cc

```diff
@@ -2316,6 +2410,70 @@ iproto_thread_init(struct iproto_thread *iproto_thread)
 	return 0;
 }
 
+static void
+tx_send_shutdown(struct cmsg *m)
+{
+	struct iproto_connection *con =
+		container_of(m, struct iproto_connection, shutdown_msg);
+	struct session *session = con->session;
+	if (session == NULL || !session->graceful_shutdown) {
+		shutdown(con->input.fd, SHUT_RD);
+	} else {
+		iproto_reply_shutdown(con->tx.p_obuf, ::schema_version);
+		if (!con->tx.is_push_sent)
+			tx_begin_push(con);
+		else
+			con->tx.is_push_pending = true;
+	}
+}
+
+static void
+iproto_process_shutdown(struct iproto_thread* iproto_thread)
+{
+	while (! rlist_empty(&iproto_thread->to_shutdown_connections)) {
+		struct iproto_connection *con =
+			rlist_first_entry(&iproto_thread->to_shutdown_connections,
+					  struct iproto_connection, to_shutdown_list);
+		rlist_del(&con->to_shutdown_list);
+		rlist_add_tail_entry(&con->iproto_thread->to_close_connections,
+				     con, to_close_list);
+		iproto_enqueue_batch(con, iproto_connection_input_buffer(con));
+		static const struct cmsg_hop send_shutdown_route[] = {
```

Comment:
It's better to declare static variable out of the loop

Reply:
Fixed